src/Markdown/Entity.elm

```diff
@@ -0,0 +1,2621 @@
+        |> Maybe.andThen String.toInt
+        |> Maybe.map validUnicode
+        |> Maybe.withDefault match.match
+
+
+validUnicode : Int -> String
+validUnicode int =
+    if isValidUnicode int && not (isBadEndUnicode int) then
+        Char.fromCode int |> String.fromChar
+
+    else
+        Char.fromCode 65533 |> String.fromChar
+
+
+
+-- Characters range from: https://www.w3.org/TR/REC-xml/#NT-Char
+
+
+isValidUnicode : Int -> Bool
+isValidUnicode int =
+    int
+        == 9
+        || int
+        == 10
+        || int
+        == 13
+        || int
+        == 133
+        || (32 <= int && int <= 126)
+        || (160 <= int && int <= 55295)
+        || (57344 <= int && int <= 64975)
+        || (65008 <= int && int <= 65533)
+        || (65536 <= int && int <= 1114109)
+
+
+
+-- Unicodes ending in FFFE and FFFF
+
+
+isBadEndUnicode : Int -> Bool
+isBadEndUnicode int =
+    let
+        remain =
+            modBy 131070 int
+
+        remain_ =
+            modBy 16 int
+    in
+    int
+        >= 131070
+        && ((0 <= remain && remain <= 15)
+                || (65536 <= remain && remain <= 65551)
+           )
+        && (remain_ == 14 || remain_ == 15)
+
+
+
+---------------------------------------------------------------------
+-- Hexadecimal numeric character
+---------------------------------------------------------------------
+
+
+replaceHexadecimals : String -> String
+replaceHexadecimals =
+    Regex.replace hexadecimalRegex replaceHexadecimal
+
+
+hexadecimalRegex : Regex
+hexadecimalRegex =
+    Regex.fromString "&#[Xx]([0-9a-fA-F]{1,8});"
+        |> Maybe.withDefault Regex.never
+
+
+replaceHexadecimal : Regex.Match -> String
+replaceHexadecimal match =
+    match.submatches
+        |> List.head
+        |> Maybe.withDefault Nothing
+        |> Maybe.map (hexToInt >> validUnicode)
+        |> Maybe.withDefault match.match
+
+
+hexToInt : String -> Int
+hexToInt =
+    String.toLower
+        >> String.toList
+        >> List.foldl
+            (\hexDigit int ->
+                int * 16 + modBy 39 (Char.toCode hexDigit) - 9
+            )
+            0
+
+
+
+---------------------------------------------------------------------
+-- Html entities references
+---------------------------------------------------------------------
+
+
+replaceEntities : String -> String
+replaceEntities =
+    Regex.replace entitiesRegex replaceEntity
+
+
+entitiesRegex : Regex
+entitiesRegex =
+    Regex.fromString "&([0-9a-zA-Z]+);"
+        |> Maybe.withDefault Regex.never
+
+
+replaceEntity : Regex.Match -> String
+replaceEntity match =
+    match.submatches
+        |> List.head
+        |> Maybe.withDefault Nothing
+        |> Maybe.andThen (\a -> Dict.get a entities)
+        |> Maybe.map (Char.fromCode >> String.fromChar)
+        |> Maybe.withDefault match.match
+
+
+
+{- Due to compiler list length issue, the list was trimmed to
+   most common entities.
+
+   More info: https://github.com/elm-lang/elm-compiler/issues/1521
```

Comment:
This issue has been solved in 0.19 so the full list of entities should be available now.

Reply:
Interesting! I also considered doing a big case statement. Any ideas on the performance and bundle size implications of doing that? In terms of generating code for that, it would be just as easy.

Reply:
That's a good question.

I would imagine that the bundle would be larger for the case statement simply because the output for the list is a JS array which is quite tight, but I'm not sure to what degree it would impact the overall package size.

As for performance, its the difference between the dictionary lookup and a js switch statement. The elm Dict has O(log n) query time. I would imagine that switch performance has to be O(n), but I don't know if the runtime can do some faster work with it or if the comparisons are cheaper for some reason.

I suppose the other performance question is around startup time. The dictionary is populated at runtime.

Reply:
Hmm, yeah it's interesting, it seems like it could compile down to something more performant on the JS side. Here's something about what you describe, the worst case O(n) performance for JS switch statements: https://www.davidbcalhoun.com/2010/is-a-hash-faster-than-a-switch-in-javascript/.

I'm wondering what the memory and bundle size impact is for having the large Dict as well. O(log n) is definitely better than O(n). I really wish Elm had the ability to do O(1) lookups like with JS objects. Seems like you could make an optimization to compile down literal Dicts to Objects. Or for that matter, any Dict. To be honest, I never really understood why Elm Dicts aren't just using JS Objects under the hood 🤷‍♂

I think we'll have to do some performance benchmarking to decide what's safe here. Let's revisit this separately 👍 Thanks for pointing this out!

Reply:
Also, it could be worth considering using a shared community resource on this, like this: https://package.elm-lang.org/packages/marcosh/elm-html-to-unicode/latest/.

Technically, markdown does have a very specific range of entities that it handles. Here's a quote from https://github.github.com/gfm/#entity-references.

> Entity references consist of & + any of the valid HTML5 entity names + ;. The document https://html.spec.whatwg.org/multipage/entities.json is used as an authoritative source for the valid entity references and their corresponding code points.

This thread describes some of the problems with this approach: https://talk.commonmark.org/t/spec-issues-character-entity-references/2306.

Another direction this could go is to give user's the ability to supply their own entity decoding function as part of the `Markdown.Renderer.Renderer`. That way, users could customize based on their performance needs, or use their preferred set of characters. But this does seem to be exposing some low-level details that would be unfortunate to clutter things up with. Just one possible direction, worth laying out all the options at least.

If we use some sort of builder style pattern, then that would be more reasonable as you could override the default rather than having to specify yourself.